src/core/namespace.c

```diff
@@ -1341,14 +1341,59 @@ static int mount_mqueuefs(const MountEntry *m) {
         return 1;
 }
 
+static int verities_find(const char *path, VeritySettings **verities, size_t *n_verities, VeritySettings **ret_verity) {
+        _cleanup_(verity_settings_done) VeritySettings verity = VERITY_SETTINGS_DEFAULT;
+        bool dup = false;
+        int r;
+
+        assert(path);
+        assert(verities);
+        assert(n_verities);
+        assert(ret_verity);
+
+        r = verity_settings_load(&verity, path, /* root_hash_path= */ NULL, /* root_hash_sig_path= */ NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to check verity root hash of %s: %m", path);
+
+        if (verity.root_hash_size == 0) {
+                *ret_verity = NULL;
+                return 1; /* 1 → this image is good, keep processing it */
+        }
+
+        FOREACH_ARRAY(v, *verities, *n_verities)
+                if (memcmp_nn(v->root_hash,
+                              v->root_hash_size,
+                              verity.root_hash,
+                              verity.root_hash_size) == 0) {
+                        dup = true;
+                        break;
+                }
```

Comment:
I'd really split out this search loop into a nice passive helper find_verity() that just searches for the root hash and returns that:

```c
static VeritySettings* find_verity(VeritySettings *haystack, size_t n_haystack, const VeritySettings *needle) {
        FOREACH_ARRAY(i, haystack, n_haystack) 
                if (memcmp_nn(…) == 0)
                        return i;
        return NULL;
}
```

you get the idea